File: src/main/java/org/stt/persistence/stt/STTItemPersister.java
package org.stt.persistence.stt;

import org.stt.model.TimeTrackingItem;
import org.stt.persistence.ItemPersister;
import org.stt.persistence.ItemReader;
import org.stt.persistence.ItemWriter;

import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;
import java.io.*;
import java.util.*;

/**
 * Writes {@link TimeTrackingItem}s to a new line. Multiline comments get joined
 * into one line: line endings \r and \n get replaced by the string \r and \n
 * respectively.
 */
@Singleton
public class STTItemPersister implements ItemPersister {

    private final STTItemConverter converter = new STTItemConverter();
    private Provider<Reader> readerProvider;
    private Provider<Writer> writerProvider;

    @Inject
    public STTItemPersister(@STTFile Provider<Reader> readerProvider,
                            @STTFile Provider<Writer> writerProvider) {
        this.readerProvider = Objects.requireNonNull(readerProvider);
        this.writerProvider = Objects.requireNonNull(writerProvider);
    }

    @Override
    public void persist(TimeTrackingItem itemToInsert) {
        Objects.requireNonNull(itemToInsert);
        StringWriter stringWriter = new StringWriter();
        Reader providedReader;
        providedReader = readerProvider.get();
        try (STTItemReader in = new STTItemReader(providedReader);
             STTItemWriter out = new STTItemWriter(stringWriter)) {
            new InsertHelper(in, out, itemToInsert).performInsert();
            rewriteFileWith(stringWriter.toString());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    public void replace(TimeTrackingItem item, TimeTrackingItem with) {
        delete(item);
        persist(with);
    }

    @Override
    public void delete(TimeTrackingItem item) {
        try (BufferedReader reader = new BufferedReader(readerProvider.get())) {
            StringWriter stringWriter = new StringWriter();
            PrintWriter printWriter = new PrintWriter(stringWriter);
            String lineOfItemToDelete = converter.timeTrackingItemToLine(item);
            String currentLine;
            while ((currentLine = reader.readLine()) != null) {
                // only persist lines which should not be deleted
                if (!currentLine.equals(lineOfItemToDelete)) {
                    printWriter.println(currentLine);
                }
            }

            printWriter.flush();
            rewriteFileWith(stringWriter.toString());
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    @Override
    public Collection<UpdatedItem> updateActivitities(Collection<TimeTrackingItem> itemsToUpdate, String newActivity) {
        try (ItemReader in = new STTItemReader(readerProvider.get());
             StringWriter sw = new StringWriter();
             ItemWriter out = new STTItemWriter(sw)) {
            Set<TimeTrackingItem> matchingItems = new HashSet<>(itemsToUpdate);
            List<UpdatedItem> updatedItems = new ArrayList<>();
            Optional<TimeTrackingItem> oItem;
            while ((oItem = in.read()).isPresent()) {
                TimeTrackingItem item = oItem.get();
                if (matchingItems.contains(item)) {
                    TimeTrackingItem updateItem = item.withActivity(newActivity);
                    updatedItems.add(new UpdatedItem(item, updateItem));
                    item = updateItem;
                }
                out.write(item);
            }
            rewriteFileWith(sw.toString());
            return updatedItems;
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private void rewriteFileWith(String content) throws IOException {
        Writer truncatingWriter = writerProvider.get();
        truncatingWriter.write(content);
        truncatingWriter.close();
    }
}

File: src/main/kotlin/org/stt/persistence/stt/STTItemPersister.kt
package org.stt.persistence.stt

import org.stt.model.TimeTrackingItem
import org.stt.persistence.ItemPersister
import java.io.*
import java.util.*
import javax.inject.Inject
import javax.inject.Provider
import javax.inject.Singleton

/**
 * Writes [TimeTrackingItem]s to a new line. Multiline comments get joined
 * into one line: line endings \r and \n get replaced by the string \r and \n
 * respectively.
 */
@Singleton
class STTItemPersister @Inject
constructor(@STTFile val readerProvider: Provider<Reader>,
            @STTFile val writerProvider: Provider<Writer>) : ItemPersister {

    private val converter = STTItemConverter()

    override fun persist(itemToInsert: TimeTrackingItem) {
        val stringWriter = StringWriter()
        val providedReader: Reader = readerProvider.get()
        try {
            STTItemReader(providedReader).use { `in` ->
                STTItemWriter(stringWriter).use { out ->
                    InsertHelper(`in`, out, itemToInsert).performInsert()
                }
                rewriteFileWith(stringWriter.toString())
            }
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }

    }

    override fun replace(item: TimeTrackingItem, with: TimeTrackingItem) {
        delete(item)
        persist(with)
    }

    override fun delete(item: TimeTrackingItem) {
        val stringWriter = StringWriter()
        val printWriter = PrintWriter(stringWriter)

        val lineOfItemToDelete = converter.timeTrackingItemToLine(item)
        BufferedReader(readerProvider.get()).useLines {
            it.filter { it != lineOfItemToDelete }.forEach { printWriter.println(it) }
        }
        rewriteFileWith(stringWriter.toString())
    }

    override fun updateActivitities(itemsToUpdate: Collection<TimeTrackingItem>, newActivity: String): Collection<ItemPersister.UpdatedItem> {
        try {
            STTItemReader(readerProvider.get()).use { `in` ->
                StringWriter().use { sw ->
                    STTItemWriter(sw).use { out ->
                        val matchingItems = HashSet(itemsToUpdate)
                        val updatedItems = ArrayList<ItemPersister.UpdatedItem>()
                        while (true) {
                            `in`.read()?.let {
                                val toWrite =
                                        if (matchingItems.contains(it)) {
                                            val updateItem = it.withActivity(newActivity)
                                            updatedItems.add(ItemPersister.UpdatedItem(it, updateItem))
                                            updateItem
                                        } else it
                                out.write(toWrite)
                            } ?: break
                        }
                        rewriteFileWith(sw.toString())
                        return updatedItems
                    }
                }
            }
        } catch (e: IOException) {
            throw UncheckedIOException(e)
        }

    }

    private fun rewriteFileWith(content: String) {
        writerProvider.get().use {
            it.write(content)
        }
    }
}
